src/Build/Logging/BinaryLogger/BuildEventArgsExtensions.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Framework;
+using static Microsoft.Build.Internal.Utilities;
+
+namespace Microsoft.Build.Logging;
+
+/// <summary>
+/// Helper extension methods for working with data passed via
+/// <see cref="ProjectEvaluationFinishedEventArgs"/> and <see cref="ProjectStartedEventArgs"/>
+/// </summary>
+public static class BuildEventArgsExtensions
+{
+    /// <summary>
+    /// Lazy enumerates and strong types properties from Properties property.
+    /// </summary>
+    public static IEnumerable<PropertyData> EnumerateProperties(
+        this ProjectEvaluationFinishedEventArgs eventArgs)
+        => EnumerateProperties(eventArgs.Properties);
+
+    /// <summary>
+    /// Lazy enumerates and strong types properties from Properties property.
+    /// </summary>
+    public static IEnumerable<PropertyData> EnumerateProperties(
+        this ProjectStartedEventArgs eventArgs)
+        => EnumerateProperties(eventArgs.Properties);
+
+    /// <summary>
+    /// Lazy enumerates and partially strong types items from Items property.
+    /// The actual item value might be wrapped to be able to provide defined interface
+    /// </summary>
+    /// <returns></returns>
+    public static IEnumerable<ItemData> EnumerateItems(
+        this ProjectEvaluationFinishedEventArgs eventArgs)
+        => EnumerateItems(eventArgs.Items);
+
+    /// <summary>
+    /// Lazy enumerates and strong types items from Items property.
+    /// The actual item value might be wrapped to be able to provide defined interface
+    /// </summary>
+    public static IEnumerable<ItemData> EnumerateItems(
+        this ProjectStartedEventArgs eventArgs)
```

Comment:
I still think we should expose a `public static IEnumerable<ItemData> EnumerateItemsOfType(this ProjectStartedEventArgs eventArgs, string typeName)` that can use the O(1) operations in `ItemDictionary`. It feels to me like "look at `@(Foo)`" is a pretty common case for an individual analysis.

Reply:
Oh - thanks for insisting and clarifying! - I completely missed that indexing part of your original suggestion.
I'll add that for sure.
It'll however guarantee O(1) only if the underlying structure supports it - which is only after creation within the worker node. After the data is transfered to main node, it changes to enumeration of `TaskItemData` - so there is no indexing possible already. But we might optimize a bit by relying on some ordering guarantee

Reply:
Done + tests